packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -505,6 +620,273 @@ class NavigatorObserver {
+  }
+
+  /// A method that will be called by the [Navigator] to decide how routes
+  /// transition in or out of the screen when [Navigator.pages] is updated.
+  ///
+  /// The `newPageRouteHistory` list contains all page-based routes in the order
+  /// that will be on the [Navigator]'s history stack after this update
+  /// completes. If a route in `newPageRouteHistory` has its
+  /// [StageableRoute.isEntering] set to true, this route requires explicit
+  /// decision on how it should transition onto the Navigator. To make a
+  /// decision, call [StageableRoute.markForPush] or [StageableRoute.markForAdd].
+  ///
+  /// The `locationToExitingPageRoute` contains the pages-based routes that
+  /// are removed from the routes history after page update and require explicit
+  /// decision on how to transition off the screen. This map records page-based
+  /// routes to be removed with the location of the route in the original route
+  /// history before the update. The keys are the locations represented by the
+  /// page-based routes that are directly below the removed routes, and the value
+  /// are the page-based routes to be removed. The location is null if the route
+  /// to be removed is the bottom most route. To make a decision for a removed
+  /// route, call [StageableRoute.markForPop], [StageableRoute.markForComplete]
+  /// or [StageableRoute.markForRemove].
+  ///
+  /// The `pageRouteToPagelessRoutes` records the page-based routes and their
+  /// associated pageless routes. If a page-based route is to be removed, its
+  /// associated pageless routes also require explicit decisions on how to
+  /// transition off the screen.
+  ///
+  /// Once all the decisions have been made, this method must merge the removed
+  /// routes and the `newPageRouteHistory` and return the merged result. The
+  /// order in the result will be the order the [Navigator] uses for updating
+  /// the route history. The return list must preserve the same order of routes
+  /// in `newPageRouteHistory`. The removed routes, however, can be inserted
+  /// into the return list freely as long as all of them are included.
+  ///
+  /// See also:
+  ///
+  ///  * [StageableRoute.markForPush], which makes route enter the screen with
+  ///    an animated transition.
+  ///  * [StageableRoute.markForAdd], which makes route enter the screen without
+  ///    an animated transition.
+  ///  * [StageableRoute.markForPop], which makes route exit the screen with
+  ///    an animated transition.
+  ///  * [StageableRoute.markForRemove], which does not complete the route and
+  ///    makes it exit the screen without an animated transition.
+  ///  * [StageableRoute.markForComplete], which completes the route and makes
+  ///    it exit the screen without an animated transition.
+  ///  * [DefaultTransitionDelegate.resolve], which implements the default way
+  ///    to decide how routes transition in or out of the screen.
+  Iterable<StageableRoute> resolve({
```

Comment:
to validate this API we should include some examples of different transition delegates that do very different things from each other. maybe enumerate some use cases for this and implement two that are on the opposite end of the spectrum.